File: src/main/java/de/fraunhofer/aisec/cpg/graph/types/PointerType.java
/*
 * Copyright (c) 2019, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.graph.types;

import java.util.Objects;
import org.neo4j.ogm.annotation.Relationship;

/**
 * PointerTypes represent all references to other Types. For C/CPP this includes pointers, as well
 * as arrays, since technically arrays are pointers. For JAVA the only use case are arrays as there
 * is no such pointer concept.
 */
public class PointerType extends Type implements SecondOrderType {

  @Relationship(value = "ELEMENT_TYPE")
  private Type elementType;

  public enum PointerOrigin {
    POINTER,
    ARRAY,
  }

  private PointerOrigin pointerOrigin;

  private PointerType() {}

  public PointerType(Type elementType, PointerOrigin pointerOrigin) {
    super();
    if (pointerOrigin == PointerOrigin.ARRAY) {
      this.setName(elementType.getName() + "[]");
      this.pointerOrigin = PointerOrigin.ARRAY;
    } else {
      this.setName(elementType.getName() + "*");
      this.pointerOrigin = PointerOrigin.POINTER;
    }
    this.elementType = elementType;
  }

  public PointerType(Type type, Type elementType, PointerOrigin pointerOrigin) {
    super(type);
    if (pointerOrigin == PointerOrigin.ARRAY) {
      this.setName(elementType.getName() + "[]");
    } else {
      this.setName(elementType.getName() + "*");
    }
    this.elementType = elementType;
    this.pointerOrigin = pointerOrigin;
  }

  /**
   * @return referencing a PointerType results in another PointerType wrapping the first
   *     PointerType, e.g. int**
   */
  @Override
  public PointerType reference(PointerOrigin origin) {
    if (origin == null) {
      origin = PointerOrigin.ARRAY;
    }
    return new PointerType(this, origin);
  }

  /** @return dereferencing a PointerType yields the type the pointer was pointing towards */
  @Override
  public Type dereference() {
    return elementType;
  }

  @Override
  public void refreshNames() {
    if (this.getElementType() instanceof PointerType) {
      ((PointerType) this.getElementType()).refreshNames();
    }
    if (this.pointerOrigin == PointerOrigin.ARRAY) {
      this.setName(this.getElementType().getName() + "[]");
    } else {
      this.setName(this.getElementType().getName() + "*");
    }
  }

  @Override
  public Type duplicate() {
    return new PointerType(this, this.elementType.duplicate(), this.pointerOrigin);
  }

  public boolean isArray() {
    return this.pointerOrigin == PointerOrigin.ARRAY;
  }

  @Override
  public boolean isSimilar(Type t) {
    if (!(t instanceof PointerType)) {
      return false;
    }

    PointerType pointerType = (PointerType) t;

    return this.getReferenceDepth() == pointerType.getReferenceDepth()
        && this.getElementType().isSimilar(pointerType.getRoot())
        && super.isSimilar(t);
  }

  public PointerOrigin getPointerOrigin() {
    return pointerOrigin;
  }

  public Type getElementType() {
    return elementType;
  }

  @Override
  public int getReferenceDepth() {
    int depth = 1;
    Type containedType = this.elementType;
    while (containedType instanceof PointerType) {
      depth++;
      containedType = ((PointerType) containedType).getElementType();
    }
    return depth;
  }

  public void setElementType(Type elementType) {
    this.elementType = elementType;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (!(o instanceof PointerType)) return false;
    if (!super.equals(o)) return false;
    PointerType that = (PointerType) o;
    return Objects.equals(elementType, that.elementType);
  }

  @Override
  public int hashCode() {
    return Objects.hash(super.hashCode(), elementType);
  }

  @Override
  public String toString() {
    return "PointerType{"
        + "elementType="
        + elementType
        + ", typeName='"
        + getName()
        + '\''
        + ", storage="
        + this.getStorage()
        + ", qualifier="
        + this.getQualifier()
        + ", origin="
        + this.getTypeOrigin()
        + '}';
  }
}

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/types/PointerType.kt
/*
 * Copyright (c) 2019, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.graph.types

import de.fraunhofer.aisec.cpg.graph.Name
import de.fraunhofer.aisec.cpg.persistence.Relationship
import java.util.*

/**
 * PointerTypes represent all references to other Types. For C/CPP this includes pointers, as well
 * as arrays, since technically arrays are pointers. For JAVA the only use case are arrays as there
 * is no such pointer concept.
 */
class PointerType : Type, SecondOrderType {
    @Relationship(value = "ELEMENT_TYPE") override var elementType: Type

    enum class PointerOrigin {
        POINTER,
        ARRAY,
    }

    var pointerOrigin: PointerOrigin? = null
        private set

    constructor(
        elementType: Type,
        pointerOrigin: PointerOrigin?,
    ) : super(
        if (pointerOrigin == PointerOrigin.ARRAY) {
            elementType.name.append("[]")
        } else {
            elementType.name.append("*")
        },
        elementType.language,
    ) {
        this.pointerOrigin = pointerOrigin
        this.elementType = elementType
        this.elementType.secondOrderTypes += this
    }

    /**
     * @return referencing a PointerType results in another PointerType wrapping the first
     *   PointerType, e.g. int**
     */
    override fun reference(pointer: PointerOrigin?): PointerType {
        var origin = pointer
        if (origin == null) {
            origin = PointerOrigin.ARRAY
        }
        return PointerType(this, origin)
    }

    /** @return dereferencing a PointerType yields the type the pointer was pointing towards */
    override fun dereference(): Type {
        return elementType
    }

    override fun refreshNames() {
        if (elementType is PointerType) {
            elementType.refreshNames()
        }
        // Use the local name (unwrap AliasType if needed for proper name)
        val effectiveElement = (elementType as? AliasType)?.underlyingType ?: elementType
        var localName = effectiveElement.name.localName
        localName +=
            if (pointerOrigin == PointerOrigin.ARRAY) {
                "[]"
            } else {
                "*"
            }
        val fullTypeName =
            Name(localName, effectiveElement.name.parent, effectiveElement.name.delimiter)
        name = fullTypeName
    }

    val isArray: Boolean
        get() = pointerOrigin == PointerOrigin.ARRAY

    override val comparisonName: String
        get() {
            // Use the underlying type's comparisonName for proper typedef comparison
            val effectiveElementType = elementType.unwrapAliasType()
            val elementName = effectiveElementType.comparisonName
            return elementName + if (pointerOrigin == PointerOrigin.ARRAY) "[]" else "*"
        }

    private fun Type.unwrapAliasType(): Type {
        var type = this
        while (type is AliasType) {
            type = type.underlyingType
        }
        return type
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        // For backward compatibility: allow AliasType to match underlying type
        val otherClass = other?.let { it::class.simpleName }
        if (otherClass == "AliasType") {
            val aliasType = other as AliasType
            // Compare using comparisonName to handle typedef aliases
            return comparisonName == aliasType.comparisonName && language == aliasType.language
        }
        if (other !is PointerType) return false
        // Compare using comparisonName to handle typedef aliases in element types
        return comparisonName == other.comparisonName && pointerOrigin == other.pointerOrigin
    }

    override fun hashCode() = Objects.hash(super.hashCode(), elementType, pointerOrigin)
}
